src/heartbeats.cpp

```diff
@@ -452,330 +299,120 @@ MegaBackupMonitor::MegaBackupMonitor(MegaClient *client)
-void MegaBackupMonitor::onSyncBackupRegistered(int syncTag, handle backupId)
-{
-    bool needsAdding = true;
-
-    // remove the tag from the set of in-flight sync registrations
-    mPendingSyncPuts.erase(syncTag);
-
-    if (ISUNDEF(backupId))
-    {
-        LOG_warn << "Received invalid id for sync with tag: " << syncTag;
-        needsAdding = false;
-    }
-
-    if (!ISUNDEF(backupId) && mHeartBeatedSyncs.find(syncTag) != mHeartBeatedSyncs.end())
-    {
-        mHeartBeatedSyncs[syncTag]->setBackupId(backupId);
-        needsAdding = false;
-    }
 
-    if (needsAdding)
-    {
-        for (const auto &hBPair : mHeartBeatedSyncs)
-        {
-            if (hBPair.second->backupId() == backupId)
-            {
-                needsAdding = false;
-                break;
-            }
-        }
-
-        if (needsAdding)
-        {
-            //create new HeartBeatSyncInfo
-            mHeartBeatedSyncs[syncTag] = std::make_shared<HeartBeatSyncInfo>(syncTag, backupId);
-        }
-    }
-
-    // store the id in the sync configuration
-    if (!ISUNDEF(backupId))
-    {
-        mClient->updateSyncBackupId(syncTag, backupId);
-    }
-
-    // handle pending updates
-    const auto &pendingSyncPair = mPendingSyncUpdates.find(syncTag);
-    if (pendingSyncPair != mPendingSyncUpdates.end())
-    {
-        if (!ISUNDEF(backupId))
-        {
-            MegaBackupInfo &info = *pendingSyncPair->second.get();
-            info.setBackupId(backupId);
-            updateBackupInfo(info);
-        }
-        else
+void MegaBackupMonitor::digestPutResult(handle backupId, UnifiedSync* syncPtr)
+{
+#ifdef ENABLE_SYNC
+    mClient->syncs.forEachUnifiedSync([&](UnifiedSync& us){
+        if (&us == syncPtr)
         {
-            LOG_warn << "discarding heartbeat update for pending sync: no valid id received for sync: " << pendingSyncPair->second->localFolder();
-
+            us.mConfig.setBackupId(backupId);
+            mClient->syncs.saveSyncConfig(us.mConfig);
         }
-        mPendingSyncUpdates.erase(pendingSyncPair);
-    }
-}
+    });
 #endif
-
-void MegaBackupMonitor::digestPutResult(handle backupId)
-{
-    if (mPendingBackupPutCallbacks.empty())
-    {
-        return;
-    }
-    // get the tag from queue of pending puts
-    auto putResultCallback = mPendingBackupPutCallbacks.front();
-    mPendingBackupPutCallbacks.pop_front();
-    //call corresponding callback
-    putResultCallback(backupId);
 }
 
-void MegaBackupMonitor::updateBackupInfo(const MegaBackupInfo &info)
+void MegaBackupMonitor::updateBackupInfo(handle backupId, const MegaBackupInfo &info)
 {
     string localFolderEncrypted(mClient->cypherTLVTextWithMasterKey("lf", info.localFolder()) );
     string deviceIdHash = mClient->getDeviceidHash();
 
-    mClient->reqs.add(new CommandBackupPut(mClient, info.backupId(), info.type(), info.megaHandle(),
+    mClient->reqs.add(new CommandBackupPut(mClient,
+                                           info.type(),
```

Comment:
This is an update, so the client needs to provide the `backupId`. Otherwise, the API will assume that it's a new registration attempt.
```suggestion
    mClient->reqs.add(new CommandBackupPut(mClient,
                                           backupdId,
                                           info.type(),
```

Reply:
excellent point!  updated